EIPS/eip-6404.md

```diff
@@ -0,0 +1,1000 @@
+
+```python
+class AccessTuple(Container):
+    address: Address
+    storage_keys: List[Hash, MAX_ACCESS_LIST_STORAGE_KEYS]
+
+class Transaction(Container):
+    chain_id: uint64  # EIP-155
+    nonce: uint64
+    max_priority_fee_per_gas: uint256  # EIP-1559
+    max_fee_per_gas: uint256  # aka `gasprice`
+    gas_limit: uint64  # aka `startgas`
+    to: Optional[Address]  # None: deploy contract
+    value: uint256
+    data: ByteList[MAX_CALLDATA_SIZE]
+    access_list: List[AccessTuple, MAX_ACCESS_LIST_SIZE]  # EIP-2930
+    blob_versioned_hashes: List[VersionedHash, MAX_VERSIONED_HASHES_LIST_SIZE]  # EIP-4844
+
+class ECDSASignature(Container):
+    y_parity: boolean  # EIP-2930
+    r: uint256
+    s: uint256
+
+class SignedTransaction(Container):
+    tx: Transaction
+    signature: ECDSASignature
+    hash_version: TxHashVersion
+
+class BlobTransactionExtension(Container):
+    blob_kzgs: List[KZGCommitment, MAX_TX_WRAP_KZG_COMMITMENTS]
+    blobs: List[Vector[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB], LIMIT_BLOBS_PER_TX]
+    kzg_aggregated_proof: KZGProof
+
+class TransactionPayload(Container):
+    signed_tx: SignedTransaction
+    blob: Optional[BlobTransactionExtension]  # EIP-4844
+```
+
+On the network, `TransactionPayload` is preceded by a single byte, [`BLOB_TX_TYPE`](./eip-4844.md#parameters) to distinguish from other transaction formats according to [EIP-2718](./eip-2718.md), regardless of the inner `signed_tx.hash_version`.
+
+The `blocks.transactions` list remains RLP encoded:
+
+```python
+block.transactions = RLP([
+    SSZ.encode(transaction_0),
+    SSZ.encode(transaction_1),
+    SSZ.encode(transaction_2),
+    ...
+])
+```
+
+#### Consensus `ExecutionPayload` changes (Transactions)
+
+The [CL `ExecutionPayload`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/eip4844/beacon-chain.md#executionpayload) is updated to replace the existing [CL `Transaction`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/bellatrix/beacon-chain.md#custom-types) container with the new `SignedTransaction` container.
+
+| Name | Value | Description |
+| - | - | - |
+| [`MAX_TRANSACTIONS_PER_PAYLOAD`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/bellatrix/beacon-chain.md#execution) | `uint64(2**20)` (= 1,048,576) | Maximum amount of transactions allowed in each block |
+
+```python
+class ExecutionPayload(Container):
+    ...
+    transactions: List[SignedTransaction, MAX_TRANSACTIONS_PER_PAYLOAD]  # Formerly, `Transaction`
+    ...
+```
+
+#### Execution block header changes (Transactions)
+
+The execution block header's `transactions_root` is updated to the `transactions_root` definition of the [consensus-specs `ExecutionPayloadHeader`](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/specs/eip4844/beacon-chain.md#executionpayloadheader).
```

Comment:
I'm not sure the reason for linking to consensus specs here. This link is also not updated with `transactions_root` as defined in this EIP. I think the link should be omitted and it should simply be stated that the two roots are now in alignment.

Reply:
The definition of `execution_payload_header.transactions_root = hash_tree_root(execution_payload.transactions)` is still the correct definition. But, yes, it can be confusing because under the link, `execution_payload.transactions` is still defined based on a different format than the proposed one. 

Removed the link and updated language. This also makes it clearer that for `withdrawals_root`, the existing format is actually reused (as opposed to `transactions_root` and `receipts_root` that change in both execution and consensus worlds).